Core/Src/drivers/nand_flash/nand_m79a_lld.c

```diff
@@ -439,40 +436,51 @@ NAND_ReturnType NAND_Block_Erase(PhysicalAddrs *addr) {
 
     /* Command 1: WRITE ENABLE */
     __write_enable();
-
+    NAND_ReturnType status = NAND_Wait_Until_Ready();
+    if (status != Ret_Success) {
+        return Ret_Failed;
+    }
     /* Command 2: BLOCK ERASE. See datasheet page 35 for details */
 
     // TODO: datasheet simply specifies block address. assuming that's the 11-bit
     // block number padded with dummy bits. check.
-    uint32_t block = 0x7ff & addr->block;
+    uint32_t block = (0x7ff & addr->block) << 6;
 
     uint8_t command[4] = {SPI_NAND_BLOCK_ERASE, BYTE_2(block), BYTE_1(block), BYTE_0(block)};
 
     SPI_Params tx_cmd = {.buffer = command, .length = 4};
     if (NAND_SPI_Send(&tx_cmd) != SPI_OK) {
-        return Ret_EraseFailed;
+        return Ret_Failed; // I/O failure
     }
 
     /* Command 3: wait for device to be ready again */
-    if (NAND_Wait_Until_Ready() != Ret_Success) {
-        return Ret_NANDBusy;
+    status = NAND_Wait_Until_Ready();
+    if (status != Ret_Success) {
+        NAND_Reset();
```

Comment:
Scary!

Reply:
How so?

Reply:
I thought NAND_Wait_Unitl_Ready() could return failure for programming and config errors. NAND_Reset() won't fix those, and the chip takes a long to become ready again once you whack it on the head.

Reply:
The only way to reset the programming and erase error flag is by issuing the reset command. You're expected to do it after a failure

Reply:
Grrr :-(